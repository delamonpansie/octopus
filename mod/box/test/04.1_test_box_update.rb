#!/usr/bin/ruby

$: << File.dirname($0) + '/lib'
require 'run_env'

class Env < RunEnv
  def config
    super + <<EOD
object_space[0].index[0].key_field[0].type = "NUM"
EOD
  end
end

Env.env_eval do
  start

  c1 = connect
  c2 = connect
  c4 = connect

  wal_writer_pid = pid + 5 # hack!
  c1.insert [3, "baz"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t1 = Thread.new { log_try { c1.insert [1, "foo"] } }
  t2 = Thread.new { sleep 0.01; log_try { c2.insert [1, "foobar"] } }

  sleep 0.05
  c4.select 1

  puts "# wal_writer cont"
  Process.kill "CONT", wal_writer_pid
  t1.join
  t2.join
  c4.select 1
  c4.delete 1
  c4.select 1


  c1.insert [3, "bar"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t2 = Thread.new { log_try { c2.update_fields 3, [0, :set, 1] } }
  t1 = Thread.new { sleep 0.01; log_try { c1.insert [1, "foo"] } }
  sleep 0.05
  c4.select 1, 3
  Process.kill "CONT", wal_writer_pid
  puts "# wal_writer cont"
  t1.join
  t2.join
  c4.select 1
end
