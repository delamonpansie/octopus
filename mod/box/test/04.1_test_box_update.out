# c1.insert([3, "baz"])
1

# wal_writer stop
# c1.insert([1, "foo"])
# c2.insert([1, "foobar"])
# c4.select(1)
[]

# wal_writer cont
1 # c1
2 # c2
# c4.select(1)
[["\x01\x00\x00\x00", "foobar"]]

# c4.delete(1)
1

# c4.select(1)
[]

# c1.insert([3, "bar"])
2

# wal_writer stop
# c2.update_fields(3, [0, :set, 1])
# c1.insert([1, "foo"])
# c4.select(1, 3)
[["\x03\x00\x00\x00", "bar"]]

# wal_writer cont
2

2

# c4.select(1)
[["\x01\x00\x00\x00", "foo"]]

